oioswift/common/middleware/container_hierarchy.py

```diff
@@ -650,28 +694,32 @@ def __call__(self, env, start_response):
             LOG.debug("%s: -> is NOT listing request", self.SWIFT_SOURCE)
             obj_parts = obj.split(self.DELIMITER)
             if len(obj_parts) > 1:
-                # path = self.DELIMITER.join(obj_parts[:-1]) + self.DELIMITER
                 path, _ = self._container_suffix(obj_parts, is_mpu)
-                is_dir = obj.endswith(self.DELIMITER)
-                # TODO (MBO) we should accept create key d1/d2/ only
-                # for empty object
+                # When Hadoop S3A detects a prefix (dir1/dir2/) for
+                # objet dir1/dir2/object1, two HEAD requests will be done:
+                # HEAD dir1/dir2/
+                #   this case is managed by checking if Redis Key OBJ exists
+                if obj.endswith(self.DELIMITER):
+                    return self._handle_dir_object(req, start_response,
+                                                   account, container, path)
+
+                # HEAD dir1/dir2 : this case is partially managed by checking is
+                # Redis Key CNT exists for dir1, if key if not present, we can
+                # reply 404 because there is no objects in dir1/ container
+                if not is_mpu and req.method in ('GET', 'HEAD'):
+                    key = self.key(account, container, CNT, path)
+                    if self.redis_keys_format == REDIS_KEYS_FORMAT_V1:
+                        res = self.conn.exists(key + '/')
+                    else:
+                        res = self.conn.hexists(key, path + "/")
+                    if not res:
+                        return self._handle_dir_object(
+                            req, start_response, account, container, path,
+                            missing=True)
```

Comment:
According the the explanation, why don't we just return `start_response("404 Not Found", [])` now?

Reply:
Good remark !